### DISCUSSION TITLE
podenv toolboxes

### DISCUSSION THREAD

**Discussion by juhp** [podenv/podenv#93] [Ideas]
What I would really like, and is missing from containers/toolbox, is being able to easily
create and run a toolbox environment (temporary and/or with/out-home, etc or not) for "any" OS image.
Currently containers/toolbox only works out of the box for very specific or custom-built images
(eg fedora-toolbox and opensuse/toolbox): eg it currently needs `mount` to be the container.

I would like to be able to do say `podenv toolbox debian` or `podenv toolbox centos` etc
and have a toolbox environment up and running immediately. It might not do
everything as well as containers/toolbox, but hopefully 60-90% say would still be pretty useful.

But this use-case seems a bit different to current podenv.
So I am wondering if it is better to try this first with some kind of a wrapper
or even as a separate tool. It probably make senses to experiment in this way anyways.

Any thoughts on this toolbox feature idea? :)

----

Also somewhat related I think it would be cool if podenv apps were more "OS modular".
eg So I could choose the app _and_ OS "backend" I want to run: eg run gimp with tumbleweed or with nix say.
Not sure how such a UI would look, but naively one could have `podenv --on nix gimp.
Further it would be cool to even be able to specify down to the OS version.
Of course it is non-trivial in that each OS has it's own package manager... so rather a lot of work.
dhall already takes quite some of way and maybe it could even do all this perhaps? I am not sure.

**Comment by TristanCacqueray** [comment]
What should be the result of `podenv toolbox debian`, e.g. what would be the expected podman command?
Note that podenv already support "lambda application", which are defined as function from Application to Application. For example see: https://github.com/podenv/hub/blob/main/Applications/nixify.dhall . So perhaps, `toolbox` could be defined similarly, then with https://github.com/podenv/podenv/pull/94 , you could use it with any image, e.g.: `podenv toolbox image:docker.io/library/debian`

Though I think it is better to at least define the package cache, for example the `fedora` application is:

```
$ podenv --show fedora
podman run -i --detach-keys "" -t --rm --hostname fedora --volume cache-dnf-latest:/var/cache/dnf --name fedora registry.fedoraproject.org/fedora:latest
```

So what is missing to make this a toolbox?

**Reply by juhp** [reply]
Okay thanks that's interesting - I will look at nixify.
The short answer to your questions is "I don't know" :-)
I guess I can starting experimenting and then I should know more - currently I expect it is mainly about doing various bind mounts and maybe the option to specify a/the home dir etc.

**Comment by TristanCacqueray** [comment]
About your second point, note that the previous versions of podenv was using such "OS modular" application.
At the very beginning, there was a `parent` system where the runtime could replace the base image: https://github.com/podenv/hub/blob/79a456d36542ed27c8b3f3e3fb22f4ada69ff2ef/config.yaml

But handling the different package manager by the runtime was tricky, thus in the next version, the application was still generic, but they were instantiated in the final set, for example see: https://github.com/podenv/hub/blob/3a529fae4b7a39f1f1e23d3975a6397f4edeecbf/defaults.dhall 

In the end, supporting multiple runtime from a single definition is more cumbersome than useful, and the current version of podenv no longer does that. Though it is still possible to make such modular application on a case by case basis, for example [firefox](https://github.com/podenv/hub/blob/main/Applications/firefox.dhall) provides two versions, the default with fusion, and the minimal without it.

Thus, right now `blender` is provided through nix, but we could add a `blender.fedora` by replacing the runtime attribute with a fedora image like so: https://github.com/podenv/hub/pull/47
